docs/dinamicko-programiranje/problem-razmjene-novca.md

```diff
@@ -0,0 +1,98 @@
+---
+title: Problem razmjene novca
+---
+
+Riječ je o problemu s kojim smo se susreli u poglavlju o pohlepnim pristupima. Pogledajmo kako bismo taj problem riješili dinamičkim programiranjem.
+
+### Rješenje
+
+Rješenje ovog problema blago se razlikuje u iterativnom i rekurzivnom pristupu. Prvo ćemo problem riješiti iterativno, a zatim rekurzivno.
+Neka je stanje neka količina novca koju smo do sad razmijenili. Stanje $DP[i]$ označava najmanji broj novčanica s kojima možemo razmijeniti količinu novca $i$. Iz tog stanja sa jednom novčanicom možemo prijeći u stanja koja su veća upravo za vrijednost neke novčanice koja nam je na raspolaganju. Krenimo od početnog stanja, u ovom zadatku zapravo imamo $k$ početnih stanja jer svaku količinu novca koja je jednaka jednoj od novčanica možemo razmijeniti upravo s tom jednom novčanicom te je vrijednost tih stanja $1$, sva ostala stanja imaju vrijednost $beskonačno$. Neka je trenutno stanje $i$, za svako stanje radimo sljedeće: povećamo $i$ za vrijednost svake novčanice i pogledamo jesmo li popravili rješenje koje smo već imali za tu(novu) vrijednost. Naravno uvijek uzimamo minimum od prethodnog i novog rješenja kako bi smo dobili optimalanu razmjenu.
+
+***Ograničenja***
+
+$x < 10000$
+
+$k < 100$
+
+$a_{i} < 10000$
+
+***ITERATIVNO***
+```cpp
+//pazimo na ovu veličinu kako tokom rješavanje ne bi izašli iz nekog polja
+const int MAXN = 20000;
+
+int x, k
+int novcanice[MAXN];
+int sol[MAXN];
+
+int main(){
+    //pretpostavimo da nam za razmjenu bilo koje količine novca treba beskonačno novčanica
+    memset(sol, 1e9, sizeof(sol));
+    cin >> k >> x;
+    for(int i = 0; i < k; i++){
+        cin >> novcanice[i];
+        //svaku količinu koja je jednaka nekoj novčanici možemo razmijeniti upravo s tom (jednom) novčanicom
+        sol[novcanice[i]] = 1;
+    }
+    for(int i = 1; i <= x; i++)
+        for(int j = 0; j < k; j++)
+            //funkcija prijelaza
+            sol[i + novcanice[j]] = min(sol[i + novcanice[j]], sol[i] + 1);
```

Comment:
mislim da ovo ima index out of range error, treba gledat jel index <=x

Reply:
Pogledaj prvi komentar u tom kodu

Reply:
sve jasno